File: api/internal/vectorize/vectorize.go
// Package vectorize parses the /fraud-score JSON payload and produces the
// 14-dimension feature vector. It deliberately avoids encoding/json — that
// path's reflection and allocation cost dominates a sub-millisecond budget.
//
// The parser is intentionally permissive: it scans for known keys within
// pre-located object/array bounds and tolerates whitespace. It does NOT
// attempt to validate the JSON's overall well-formedness; malformed input
// returns an error and the server responds 400.
package vectorize

import (
	"bytes"
	"errors"
	"strconv"
	"unsafe"

	"github.com/rinha2026/sketchy/api/internal/mcc"
)

// Constants from resources/normalization.json. These never change during the
// edition, so we compile them in.
const (
	maxAmount             = 10000.0
	maxInstallments       = 12.0
	amountVsAvgRatio      = 10.0
	maxMinutes            = 1440.0
	maxKM                 = 1000.0
	maxTxCount24h         = 20.0
	maxMerchantAvgAmount  = 10000.0
)

// ErrParse is returned when the JSON payload is missing required fields or
// has malformed values. Callers map this to HTTP 400.
var ErrParse = errors.New("vectorize: malformed payload")

// Build parses body and writes 14 normalized features into q.
//
// Index meaning (from DETECTION_RULES.md):
//
//	0  amount                clamp(amount / 10000)
//	1  installments          clamp(installments / 12)
//	2  amount_vs_avg         clamp((amount / customer.avg_amount) / 10)
//	3  hour_of_day           hour / 23 (UTC)
//	4  day_of_week           weekday / 6 (Mon=0..Sun=6)
//	5  minutes_since_last_tx clamp(min / 1440) OR -1 if last_transaction is null
//	6  km_from_last_tx       clamp(km_from_current / 1000) OR -1 if null
//	7  km_from_home          clamp(km_from_home / 1000)
//	8  tx_count_24h          clamp(tx_count_24h / 20)
//	9  is_online             1.0 / 0.0
//	10 card_present          1.0 / 0.0
//	11 unknown_merchant      1.0 if merchant.id NOT in customer.known_merchants
//	12 mcc_risk              from mcc package
//	13 merchant_avg_amount   clamp(merchant.avg_amount / 10000)
func Build(body []byte, q *[14]float32) error {
	txStart, txEnd, ok := objectRange(body, 0, len(body), keyTransaction)
	if !ok {
		return ErrParse
	}
	custStart, custEnd, ok := objectRange(body, 0, len(body), keyCustomer)
	if !ok {
		return ErrParse
	}
	merchStart, merchEnd, ok := objectRange(body, 0, len(body), keyMerchant)
	if !ok {
		return ErrParse
	}
	termStart, termEnd, ok := objectRange(body, 0, len(body), keyTerminal)
	if !ok {
		return ErrParse
	}

	amount, ok := jsonNumber(body, txStart, txEnd, keyAmount)
	if !ok {
		return ErrParse
	}
	installments, ok := jsonNumber(body, txStart, txEnd, keyInstallments)
	if !ok {
		return ErrParse
	}
	requestedAt, ok := jsonString(body, txStart, txEnd, keyRequestedAt)
	if !ok || len(requestedAt) < 19 {
		return ErrParse
	}

	customerAvg, ok := jsonNumber(body, custStart, custEnd, keyAvgAmount)
	if !ok {
		return ErrParse
	}
	tx24h, ok := jsonNumber(body, custStart, custEnd, keyTxCount24h)
	if !ok {
		return ErrParse
	}

	merchantID, ok := jsonString(body, merchStart, merchEnd, keyID)
	if !ok {
		return ErrParse
	}
	mccBytes, ok := jsonString(body, merchStart, merchEnd, keyMCC)
	if !ok {
		return ErrParse
	}
	merchantAvg, ok := jsonNumber(body, merchStart, merchEnd, keyAvgAmount)
	if !ok {
		return ErrParse
	}

	isOnline, ok := jsonBool(body, termStart, termEnd, keyIsOnline)
	if !ok {
		return ErrParse
	}
	cardPresent, ok := jsonBool(body, termStart, termEnd, keyCardPresent)
	if !ok {
		return ErrParse
	}
	kmFromHome, ok := jsonNumber(body, termStart, termEnd, keyKMFromHome)
	if !ok {
		return ErrParse
	}

	// last_transaction is optional and may be null.
	minutesSinceLast := float32(-1)
	kmFromLast := float32(-1)
	if ltStart, ltEnd, isObj := objectRange(body, 0, len(body), keyLastTransaction); isObj {
		lastTS, ok := jsonString(body, ltStart, ltEnd, keyTimestamp)
		if !ok || len(lastTS) < 19 {
			return ErrParse
		}
		kmFromCurrent, ok := jsonNumber(body, ltStart, ltEnd, keyKMFromCurrent)
		if !ok {
			return ErrParse
		}
		mins := minutesBetween(requestedAt, lastTS)
		minutesSinceLast = clamp01(float32(mins) / maxMinutes)
		kmFromLast = clamp01(kmFromCurrent / maxKM)
	}

	// unknown_merchant: 1.0 if merchant.id NOT present in customer.known_merchants.
	known := arrayContainsString(body, custStart, custEnd, keyKnownMerchants, merchantID)
	unknownMerchant := float32(1)
	if known {
		unknownMerchant = 0
	}

	amountVsAvg := float32(1)
	if customerAvg > 0 {
		amountVsAvg = clamp01((amount / customerAvg) / amountVsAvgRatio)
	}

	q[0] = clamp01(amount / maxAmount)
	q[1] = clamp01(installments / maxInstallments)
	q[2] = amountVsAvg
	q[3] = clamp01(float32(isoHourUTC(requestedAt)) / 23.0)
	q[4] = clamp01(float32(weekdayMondayZero(requestedAt)) / 6.0)
	q[5] = minutesSinceLast
	q[6] = kmFromLast
	q[7] = clamp01(kmFromHome / maxKM)
	q[8] = clamp01(tx24h / maxTxCount24h)
	if isOnline {
		q[9] = 1
	} else {
		q[9] = 0
	}
	if cardPresent {
		q[10] = 1
	} else {
		q[10] = 0
	}
	q[11] = unknownMerchant
	q[12] = mcc.Get(mccBytes)
	q[13] = clamp01(merchantAvg / maxMerchantAvgAmount)
	return nil
}

// Pre-quoted key bytes. Allocating these once saves both the alloc and the
// implicit "key" → []byte conversion on every request.
var (
	keyTransaction     = []byte(`"transaction"`)
	keyCustomer        = []byte(`"customer"`)
	keyMerchant        = []byte(`"merchant"`)
	keyTerminal        = []byte(`"terminal"`)
	keyLastTransaction = []byte(`"last_transaction"`)
	keyAmount          = []byte(`"amount"`)
	keyInstallments    = []byte(`"installments"`)
	keyRequestedAt     = []byte(`"requested_at"`)
	keyAvgAmount       = []byte(`"avg_amount"`)
	keyTxCount24h      = []byte(`"tx_count_24h"`)
	keyKnownMerchants  = []byte(`"known_merchants"`)
	keyID              = []byte(`"id"`)
	keyMCC             = []byte(`"mcc"`)
	keyIsOnline        = []byte(`"is_online"`)
	keyCardPresent     = []byte(`"card_present"`)
	keyKMFromHome      = []byte(`"km_from_home"`)
	keyTimestamp       = []byte(`"timestamp"`)
	keyKMFromCurrent   = []byte(`"km_from_current"`)
)

// findKey locates the first occurrence of key (already including its quotes,
// e.g. `"amount"`) in body[start:end] and returns the absolute offset.
// Returns -1 if not found.
func findKey(body []byte, start, end int, key []byte) int {
	if start >= end {
		return -1
	}
	idx := bytes.Index(body[start:end], key)
	if idx < 0 {
		return -1
	}
	return start + idx
}

// objectRange locates "key": { … } and returns [absolute_lbrace, absolute_rbrace_after_close).
// Returns ok=false if the value isn't an object (e.g. is null).
func objectRange(body []byte, start, end int, key []byte) (int, int, bool) {
	k := findKey(body, start, end, key)
	if k < 0 {
		return 0, 0, false
	}
	colon := indexByteFrom(body, k+len(key), end, ':')
	if colon < 0 {
		return 0, 0, false
	}
	p := skipWS(body, colon+1, end)
	if p >= end || body[p] != '{' {
		return 0, 0, false
	}
	close := matchBrace(body, p, end)
	if close < 0 {
		return 0, 0, false
	}
	return p, close, true
}

// jsonNumber parses key's numeric value as float32. Returns ok=false on missing key
// or unparseable number.
func jsonNumber(body []byte, start, end int, key []byte) (float32, bool) {
	k := findKey(body, start, end, key)
	if k < 0 {
		return 0, false
	}
	colon := indexByteFrom(body, k+len(key), end, ':')
	if colon < 0 {
		return 0, false
	}
	p := skipWS(body, colon+1, end)
	if p >= end {
		return 0, false
	}
	// Find the end of the number — a comma, brace, bracket, or whitespace.
	q := p
	if body[q] == '-' || body[q] == '+' {
		q++
	}
	for q < end {
		c := body[q]
		if (c >= '0' && c <= '9') || c == '.' || c == 'e' || c == 'E' || c == '-' || c == '+' {
			q++
			continue
		}
		break
	}
	if q == p {
		return 0, false
	}
	// unsafe.String avoids the string allocation that `string(body[p:q])` would
	// trigger on every numeric field (~10 per request). The body buffer is not
	// mutated during ParseFloat, and ParseFloat does not retain the string.
	v, err := strconv.ParseFloat(unsafe.String(&body[p], q-p), 32)
	if err != nil {
		return 0, false
	}
	return float32(v), true
}

func jsonBool(body []byte, start, end int, key []byte) (bool, bool) {
	k := findKey(body, start, end, key)
	if k < 0 {
		return false, false
	}
	colon := indexByteFrom(body, k+len(key), end, ':')
	if colon < 0 {
		return false, false
	}
	p := skipWS(body, colon+1, end)
	if p+4 <= end && body[p] == 't' && body[p+1] == 'r' && body[p+2] == 'u' && body[p+3] == 'e' {
		return true, true
	}
	if p+5 <= end && body[p] == 'f' && body[p+1] == 'a' && body[p+2] == 'l' && body[p+3] == 's' && body[p+4] == 'e' {
		return false, true
	}
	return false, false
}

// jsonString returns the inner-quoted bytes (no copy) for key's string value.
// Caller must not mutate the returned slice. Returns ok=false on missing key
// or non-string value.
func jsonString(body []byte, start, end int, key []byte) ([]byte, bool) {
	k := findKey(body, start, end, key)
	if k < 0 {
		return nil, false
	}
	colon := indexByteFrom(body, k+len(key), end, ':')
	if colon < 0 {
		return nil, false
	}
	p := skipWS(body, colon+1, end)
	if p >= end || body[p] != '"' {
		return nil, false
	}
	p++
	q := p
	for q < end && body[q] != '"' {
		q++
	}
	if q >= end {
		return nil, false
	}
	return body[p:q], true
}

// arrayContainsString returns true if key's array (a JSON list of strings)
// contains needle (compared by raw bytes).
func arrayContainsString(body []byte, start, end int, key, needle []byte) bool {
	k := findKey(body, start, end, key)
	if k < 0 {
		return false
	}
	colon := indexByteFrom(body, k+len(key), end, ':')
	if colon < 0 {
		return false
	}
	lb := indexByteFrom(body, colon+1, end, '[')
	if lb < 0 {
		return false
	}
	p := lb + 1
	for p < end {
		p = skipWS(body, p, end)
		if p >= end || body[p] == ']' {
			return false
		}
		if body[p] != '"' {
			p++
			continue
		}
		p++
		q := p
		for q < end && body[q] != '"' {
			q++
		}
		if q >= end {
			return false
		}
		if bytes.Equal(body[p:q], needle) {
			return true
		}
		p = q + 1
	}
	return false
}

func indexByteFrom(body []byte, start, end int, c byte) int {
	if start >= end {
		return -1
	}
	idx := bytes.IndexByte(body[start:end], c)
	if idx < 0 {
		return -1
	}
	return start + idx
}

func skipWS(body []byte, start, end int) int {
	for start < end {
		c := body[start]
		if c == ' ' || c == '\t' || c == '\n' || c == '\r' {
			start++
			continue
		}
		return start
	}
	return end
}

// matchBrace assumes body[start] == '{' and returns the absolute index AFTER
// the matching '}'. Skips quoted strings and escaped chars.
func matchBrace(body []byte, start, end int) int {
	depth := 0
	inStr := false
	esc := false
	for p := start; p < end; p++ {
		c := body[p]
		if inStr {
			if esc {
				esc = false
				continue
			}
			if c == '\\' {
				esc = true
				continue
			}
			if c == '"' {
				inStr = false
			}
			continue
		}
		switch c {
		case '"':
			inStr = true
		case '{':
			depth++
		case '}':
			depth--
			if depth == 0 {
				return p + 1
			}
		}
	}
	return -1
}

func clamp01(x float32) float32 {
	if x < 0 {
		return 0
	}
	if x > 1 {
		return 1
	}
	return x
}

// isoHourUTC parses positions 11-12 of an ISO-8601 timestamp (e.g. "2026-03-11T20:23:35Z").
// Bounds-checked at the caller (timestamp is at least 19 chars).
func isoHourUTC(ts []byte) int {
	h := int(ts[11]-'0')*10 + int(ts[12]-'0')
	if h < 0 {
		return 0
	}
	if h > 23 {
		return 23
	}
	return h
}

// weekdayMondayZero returns 0 (Mon) .. 6 (Sun) for the date in ts.
// Uses Howard Hinnant's "days from civil" algorithm — pure integer math.
func weekdayMondayZero(ts []byte) int {
	y := int(ts[0]-'0')*1000 + int(ts[1]-'0')*100 + int(ts[2]-'0')*10 + int(ts[3]-'0')
	m := int(ts[5]-'0')*10 + int(ts[6]-'0')
	d := int(ts[8]-'0')*10 + int(ts[9]-'0')
	days := daysFromCivil(y, m, d)
	w := int((days + 3) % 7) // 1970-01-01 was a Thursday (=3 in mon-zero scheme)
	if w < 0 {
		w += 7
	}
	return w
}

// minutesBetween returns |a - b| in minutes. Both must be ISO-8601 to-second.
func minutesBetween(a, b []byte) int64 {
	sa := isoToEpochSeconds(a)
	sb := isoToEpochSeconds(b)
	d := sa - sb
	if d < 0 {
		d = -d
	}
	return d / 60
}

func isoToEpochSeconds(ts []byte) int64 {
	y := int(ts[0]-'0')*1000 + int(ts[1]-'0')*100 + int(ts[2]-'0')*10 + int(ts[3]-'0')
	m := int(ts[5]-'0')*10 + int(ts[6]-'0')
	d := int(ts[8]-'0')*10 + int(ts[9]-'0')
	hh := int(ts[11]-'0')*10 + int(ts[12]-'0')
	mm := int(ts[14]-'0')*10 + int(ts[15]-'0')
	ss := int(ts[17]-'0')*10 + int(ts[18]-'0')
	days := daysFromCivil(y, m, d)
	return days*86400 + int64(hh)*3600 + int64(mm)*60 + int64(ss)
}

// daysFromCivil — Howard Hinnant's algorithm. Returns days since 1970-01-01
// for a Gregorian (year, month, day). Works for all valid inputs.
func daysFromCivil(y, m, d int) int64 {
	if m <= 2 {
		y--
	}
	var era int
	if y >= 0 {
		era = y / 400
	} else {
		era = (y - 399) / 400
	}
	yoe := uint(y - era*400)
	var monthShift int
	if m > 2 {
		monthShift = -3
	} else {
		monthShift = 9
	}
	doy := uint((153*(uint(m)+uint(monthShift))+2)/5) + uint(d) - 1
	doe := yoe*365 + yoe/4 - yoe/100 + doy
	return int64(era)*146097 + int64(doe) - 719468
}

File: api-rust/src/vectorize.rs
// Manual JSON parser → 14-dim feature vector. Direct port of
// api/internal/vectorize/vectorize.go. Same key names, same normalization,
// same date arithmetic — must produce byte-identical f32 output to the Go
// version (otherwise the IVF top-5 diverges and detection drops).

use crate::mcc;

const MAX_AMOUNT: f32 = 10000.0;
const MAX_INSTALLMENTS: f32 = 12.0;
const AMOUNT_VS_AVG_RATIO: f32 = 10.0;
const MAX_MINUTES: f32 = 1440.0;
const MAX_KM: f32 = 1000.0;
const MAX_TX_COUNT_24H: f32 = 20.0;
const MAX_MERCHANT_AVG_AMOUNT: f32 = 10000.0;

const KEY_TRANSACTION: &[u8] = b"\"transaction\"";
const KEY_CUSTOMER: &[u8] = b"\"customer\"";
const KEY_MERCHANT: &[u8] = b"\"merchant\"";
const KEY_TERMINAL: &[u8] = b"\"terminal\"";
const KEY_LAST_TRANSACTION: &[u8] = b"\"last_transaction\"";
const KEY_AMOUNT: &[u8] = b"\"amount\"";
const KEY_INSTALLMENTS: &[u8] = b"\"installments\"";
const KEY_REQUESTED_AT: &[u8] = b"\"requested_at\"";
const KEY_AVG_AMOUNT: &[u8] = b"\"avg_amount\"";
const KEY_TX_COUNT_24H: &[u8] = b"\"tx_count_24h\"";
const KEY_KNOWN_MERCHANTS: &[u8] = b"\"known_merchants\"";
const KEY_ID: &[u8] = b"\"id\"";
const KEY_MCC: &[u8] = b"\"mcc\"";
const KEY_IS_ONLINE: &[u8] = b"\"is_online\"";
const KEY_CARD_PRESENT: &[u8] = b"\"card_present\"";
const KEY_KM_FROM_HOME: &[u8] = b"\"km_from_home\"";
const KEY_TIMESTAMP: &[u8] = b"\"timestamp\"";
const KEY_KM_FROM_CURRENT: &[u8] = b"\"km_from_current\"";

#[derive(Debug)]
pub struct ParseError;

pub fn build(body: &[u8], q: &mut [f32; 14], mcc_table: &mcc::Table) -> Result<(), ParseError> {
    let (tx_start, tx_end) = object_range(body, 0, body.len(), KEY_TRANSACTION).ok_or(ParseError)?;
    let (cust_start, cust_end) =
        object_range(body, 0, body.len(), KEY_CUSTOMER).ok_or(ParseError)?;
    let (merch_start, merch_end) =
        object_range(body, 0, body.len(), KEY_MERCHANT).ok_or(ParseError)?;
    let (term_start, term_end) =
        object_range(body, 0, body.len(), KEY_TERMINAL).ok_or(ParseError)?;

    let amount = json_number(body, tx_start, tx_end, KEY_AMOUNT).ok_or(ParseError)?;
    let installments = json_number(body, tx_start, tx_end, KEY_INSTALLMENTS).ok_or(ParseError)?;
    let requested_at = json_string(body, tx_start, tx_end, KEY_REQUESTED_AT).ok_or(ParseError)?;
    if requested_at.len() < 19 {
        return Err(ParseError);
    }

    let customer_avg = json_number(body, cust_start, cust_end, KEY_AVG_AMOUNT).ok_or(ParseError)?;
    let tx24h = json_number(body, cust_start, cust_end, KEY_TX_COUNT_24H).ok_or(ParseError)?;

    let merchant_id = json_string(body, merch_start, merch_end, KEY_ID).ok_or(ParseError)?;
    let mcc_bytes = json_string(body, merch_start, merch_end, KEY_MCC).ok_or(ParseError)?;
    let merchant_avg =
        json_number(body, merch_start, merch_end, KEY_AVG_AMOUNT).ok_or(ParseError)?;

    let is_online = json_bool(body, term_start, term_end, KEY_IS_ONLINE).ok_or(ParseError)?;
    let card_present = json_bool(body, term_start, term_end, KEY_CARD_PRESENT).ok_or(ParseError)?;
    let km_from_home =
        json_number(body, term_start, term_end, KEY_KM_FROM_HOME).ok_or(ParseError)?;

    let mut minutes_since_last: f32 = -1.0;
    let mut km_from_last: f32 = -1.0;
    if let Some((lt_start, lt_end)) = object_range(body, 0, body.len(), KEY_LAST_TRANSACTION) {
        let last_ts = json_string(body, lt_start, lt_end, KEY_TIMESTAMP).ok_or(ParseError)?;
        if last_ts.len() < 19 {
            return Err(ParseError);
        }
        let km_from_current =
            json_number(body, lt_start, lt_end, KEY_KM_FROM_CURRENT).ok_or(ParseError)?;
        let mins = minutes_between(requested_at, last_ts);
        minutes_since_last = clamp01(mins as f32 / MAX_MINUTES);
        km_from_last = clamp01(km_from_current / MAX_KM);
    }

    let known = array_contains_string(body, cust_start, cust_end, KEY_KNOWN_MERCHANTS, merchant_id);
    let unknown_merchant: f32 = if known { 0.0 } else { 1.0 };

    let amount_vs_avg = if customer_avg > 0.0 {
        clamp01((amount / customer_avg) / AMOUNT_VS_AVG_RATIO)
    } else {
        1.0
    };

    q[0] = clamp01(amount / MAX_AMOUNT);
    q[1] = clamp01(installments / MAX_INSTALLMENTS);
    q[2] = amount_vs_avg;
    q[3] = clamp01(iso_hour_utc(requested_at) as f32 / 23.0);
    q[4] = clamp01(weekday_monday_zero(requested_at) as f32 / 6.0);
    q[5] = minutes_since_last;
    q[6] = km_from_last;
    q[7] = clamp01(km_from_home / MAX_KM);
    q[8] = clamp01(tx24h / MAX_TX_COUNT_24H);
    q[9] = if is_online { 1.0 } else { 0.0 };
    q[10] = if card_present { 1.0 } else { 0.0 };
    q[11] = unknown_merchant;
    q[12] = mcc_table.get(mcc_bytes);
    q[13] = clamp01(merchant_avg / MAX_MERCHANT_AVG_AMOUNT);
    Ok(())
}

#[inline]
fn clamp01(x: f32) -> f32 {
    if x < 0.0 {
        0.0
    } else if x > 1.0 {
        1.0
    } else {
        x
    }
}

fn find_key(body: &[u8], start: usize, end: usize, key: &[u8]) -> Option<usize> {
    if start >= end {
        return None;
    }
    memmem(&body[start..end], key).map(|p| start + p)
}

fn memmem(haystack: &[u8], needle: &[u8]) -> Option<usize> {
    if needle.is_empty() {
        return Some(0);
    }
    if needle.len() > haystack.len() {
        return None;
    }
    let last = haystack.len() - needle.len();
    let first = needle[0];
    let mut i = 0;
    while i <= last {
        if haystack[i] == first && &haystack[i..i + needle.len()] == needle {
            return Some(i);
        }
        i += 1;
    }
    None
}

fn index_byte_from(body: &[u8], start: usize, end: usize, c: u8) -> Option<usize> {
    if start >= end {
        return None;
    }
    body[start..end].iter().position(|&b| b == c).map(|p| start + p)
}

fn skip_ws(body: &[u8], mut start: usize, end: usize) -> usize {
    while start < end {
        match body[start] {
            b' ' | b'\t' | b'\n' | b'\r' => start += 1,
            _ => return start,
        }
    }
    end
}

fn match_brace(body: &[u8], start: usize, end: usize) -> Option<usize> {
    let mut depth: i32 = 0;
    let mut in_str = false;
    let mut esc = false;
    let mut p = start;
    while p < end {
        let c = body[p];
        if in_str {
            if esc {
                esc = false;
            } else if c == b'\\' {
                esc = true;
            } else if c == b'"' {
                in_str = false;
            }
            p += 1;
            continue;
        }
        match c {
            b'"' => in_str = true,
            b'{' => depth += 1,
            b'}' => {
                depth -= 1;
                if depth == 0 {
                    return Some(p + 1);
                }
            }
            _ => {}
        }
        p += 1;
    }
    None
}

fn object_range(body: &[u8], start: usize, end: usize, key: &[u8]) -> Option<(usize, usize)> {
    let k = find_key(body, start, end, key)?;
    let colon = index_byte_from(body, k + key.len(), end, b':')?;
    let p = skip_ws(body, colon + 1, end);
    if p >= end || body[p] != b'{' {
        return None;
    }
    let close = match_brace(body, p, end)?;
    Some((p, close))
}

fn json_number(body: &[u8], start: usize, end: usize, key: &[u8]) -> Option<f32> {
    let k = find_key(body, start, end, key)?;
    let colon = index_byte_from(body, k + key.len(), end, b':')?;
    let p = skip_ws(body, colon + 1, end);
    if p >= end {
        return None;
    }
    let mut q = p;
    if body[q] == b'-' || body[q] == b'+' {
        q += 1;
    }
    while q < end {
        let c = body[q];
        let digit_or_sign = c.is_ascii_digit()
            || c == b'.'
            || c == b'e'
            || c == b'E'
            || c == b'-'
            || c == b'+';
        if !digit_or_sign {
            break;
        }
        q += 1;
    }
    if q == p {
        return None;
    }
    let raw = std::str::from_utf8(&body[p..q]).ok()?;
    raw.parse::<f32>().ok()
}

fn json_bool(body: &[u8], start: usize, end: usize, key: &[u8]) -> Option<bool> {
    let k = find_key(body, start, end, key)?;
    let colon = index_byte_from(body, k + key.len(), end, b':')?;
    let p = skip_ws(body, colon + 1, end);
    if p + 4 <= end && &body[p..p + 4] == b"true" {
        return Some(true);
    }
    if p + 5 <= end && &body[p..p + 5] == b"false" {
        return Some(false);
    }
    None
}

fn json_string<'a>(body: &'a [u8], start: usize, end: usize, key: &[u8]) -> Option<&'a [u8]> {
    let k = find_key(body, start, end, key)?;
    let colon = index_byte_from(body, k + key.len(), end, b':')?;
    let p = skip_ws(body, colon + 1, end);
    if p >= end || body[p] != b'"' {
        return None;
    }
    let p = p + 1;
    let mut q = p;
    while q < end && body[q] != b'"' {
        q += 1;
    }
    if q >= end {
        return None;
    }
    Some(&body[p..q])
}

fn array_contains_string(
    body: &[u8],
    start: usize,
    end: usize,
    key: &[u8],
    needle: &[u8],
) -> bool {
    let Some(k) = find_key(body, start, end, key) else {
        return false;
    };
    let Some(colon) = index_byte_from(body, k + key.len(), end, b':') else {
        return false;
    };
    let Some(lb) = index_byte_from(body, colon + 1, end, b'[') else {
        return false;
    };
    let mut p = lb + 1;
    while p < end {
        p = skip_ws(body, p, end);
        if p >= end || body[p] == b']' {
            return false;
        }
        if body[p] != b'"' {
            p += 1;
            continue;
        }
        p += 1;
        let mut q = p;
        while q < end && body[q] != b'"' {
            q += 1;
        }
        if q >= end {
            return false;
        }
        if &body[p..q] == needle {
            return true;
        }
        p = q + 1;
    }
    false
}

fn iso_hour_utc(ts: &[u8]) -> i32 {
    let h = (ts[11] - b'0') as i32 * 10 + (ts[12] - b'0') as i32;
    if h < 0 {
        0
    } else if h > 23 {
        23
    } else {
        h
    }
}

fn weekday_monday_zero(ts: &[u8]) -> i32 {
    let y = (ts[0] - b'0') as i32 * 1000
        + (ts[1] - b'0') as i32 * 100
        + (ts[2] - b'0') as i32 * 10
        + (ts[3] - b'0') as i32;
    let m = (ts[5] - b'0') as i32 * 10 + (ts[6] - b'0') as i32;
    let d = (ts[8] - b'0') as i32 * 10 + (ts[9] - b'0') as i32;
    let days = days_from_civil(y, m, d);
    let mut w = ((days + 3) % 7) as i32;
    if w < 0 {
        w += 7;
    }
    w
}

fn minutes_between(a: &[u8], b: &[u8]) -> i64 {
    let sa = iso_to_epoch_seconds(a);
    let sb = iso_to_epoch_seconds(b);
    let mut d = sa - sb;
    if d < 0 {
        d = -d;
    }
    d / 60
}

fn iso_to_epoch_seconds(ts: &[u8]) -> i64 {
    let y = (ts[0] - b'0') as i32 * 1000
        + (ts[1] - b'0') as i32 * 100
        + (ts[2] - b'0') as i32 * 10
        + (ts[3] - b'0') as i32;
    let m = (ts[5] - b'0') as i32 * 10 + (ts[6] - b'0') as i32;
    let d = (ts[8] - b'0') as i32 * 10 + (ts[9] - b'0') as i32;
    let hh = (ts[11] - b'0') as i64 * 10 + (ts[12] - b'0') as i64;
    let mm = (ts[14] - b'0') as i64 * 10 + (ts[15] - b'0') as i64;
    let ss = (ts[17] - b'0') as i64 * 10 + (ts[18] - b'0') as i64;
    let days = days_from_civil(y, m, d);
    days * 86400 + hh * 3600 + mm * 60 + ss
}

// Howard Hinnant's "days from civil" — pure integer date math, identical to
// the Go implementation in vectorize.go.
fn days_from_civil(mut y: i32, m: i32, d: i32) -> i64 {
    if m <= 2 {
        y -= 1;
    }
    let era = if y >= 0 { y / 400 } else { (y - 399) / 400 };
    let yoe = (y - era * 400) as i64; // 0..=399
    let month_shift: i64 = if m > 2 { -3 } else { 9 };
    let doy = (153 * (m as i64 + month_shift) + 2) / 5 + d as i64 - 1;
    let doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;
    era as i64 * 146097 + doe - 719468
}
